File: problems-originals/0101-0200/0189_rotate-array/solution_cyclic_replacement.go
// The positions split into gcd(n, k) cycles under i -> (i + k) % n — walking
// each cycle carries its values straight to their final slots with only one
// element in flight at a time, so the rotation rewrites the given slice with
// no second allocation.
func rotate(nums []int, k int) []int {
	n := len(nums)
	// A rotation by n steps is the identity, so any larger k wraps
	// around to k % n — normalize before chasing cycles.
	k %= n
	gcd := func(a, b int) int {
		for b != 0 {
			a, b = b, a%b
		}
		return a
	}
	cycles := gcd(n, k)
	for start := 0; start < cycles; start++ {
		carried := nums[start]
		j := start
		for {
			// Drop the carried element into its rightful slot and catch
			// the one displaced; the cycle closes back at the start.
			nxt := (j + k) % n
			nums[nxt], carried = carried, nums[nxt]
			j = nxt
			if nxt == start {
				break
			}
		}
	}
	// The rotation happened inside the input allocation; the same slice,
	// now rotated, is what the judge compares.
	return nums
}

File: problems-adapt/0101-0200/0189_roll-the-array-rightward/solution_cyclic_replacement.rs
impl Solution {
    pub fn roll_right(mut nums: Vec<i32>, k: i32) -> Vec<i32> {
        // Ownership hands over the whole allocation, so the cycle walks
        // rewrite it in place — no second array is ever built.
        let n = nums.len();
        // A rotation by n steps is the identity, so any larger k wraps
        // around to k % n — normalize before chasing cycles.
        let k = (k as usize) % n;
        // The positions split into gcd(n, k) cycles under i -> (i + k) % n,
        // and walking each cycle carries its values straight to their final
        // slots with only one element in flight at a time. Euclid's count:
        // gcd(n, 0) is n, so a fully-normalized k of zero collapses into n
        // one-position cycles that change nothing.
        let mut cycles = n;
        let mut rest = k;
        while rest != 0 {
            let temp = cycles % rest;
            cycles = rest;
            rest = temp;
        }
        for start in 0..cycles {
            let mut carried = nums[start];
            let mut j = start;
            loop {
                // Drop the carried element into its rightful slot and catch
                // the one displaced; the cycle closes back at the start.
                let nxt = (j + k) % n;
                let picked = nums[nxt];
                nums[nxt] = carried;
                carried = picked;
                j = nxt;
                if nxt == start {
                    break;
                }
            }
        }
        nums
    }
}
